### ISSUE TITLE
XM Directory functions

### ISSUE THREAD

**Issue by rywood89** [ropensci/qualtRics#131]
Hi,

I have enjoyed working with ```qualtRics``` and was glad to see that it's being updated! Thank you.

I'm working with Qualtrics' XM Directory, which adds another layer of [API documentation](https://api.qualtrics.com/docs/transactions-in-iq-directory) to access and edit XM Directory contacts and mailing lists. I've been able to use ```httr```, ```jsontlite``` and Qualtrics' API calls to do what I need to do, but I'm wondering if you plan on adding some XM Directory functions similar to what the Python library [```QualtricsAPI```](https://github.com/Jaseibert/QualtricsAPI/tree/master/QualtricsAPI/XM) has done?

I'd like to offer help in translating from the Python library if I can, but I've just started learning Python. Also, I'm learning GitHub. (I literally signed up to make this request.)

Thanks again for your efforts.

**Comment by juliasilge** [comment]
Thank you so much for this issue @rywood89! Would you be willing to share the code you wrote using **httr** and **jsonlite** to get to the XM Directory contacts and mailing lists? I am not sure if the Qualtrics account I use has access to this feature so I'll definitely need some help getting this into the package. We could start with you just showing me what kind of thing you are doing now, if that works for you.

**Comment by rywood89** [comment]
Great. I'm happy to help anyway I can, but I'm a novice when it comes to many things.
For example, in successfully making my life difficult by avoiding a solid understanding of loops, I wrote some "manual" code that pulls the pages of data from the API call. After I pulled the pages, extracted the elements, and cleaned the formatting, I bound things together. 

```
library('tidyverse')
library('httr')
library('jsonlite')
library("beepr")
my_API_token <- <API_token_from_Qualtrics>
baseurl <- "https://<organization_ID>.qualtrics.com/API/v3/directories/<POOL_id>/contacts"

## Add in API_token_from_Qualtrics, organization_ID, and POOL_id for your own XM Directory.

##### Get first page of 500
first_page_get <- GET(paste0(baseurl,"?pageSize=500&useNewPaginationScheme=true"), add_headers("x-api-token" = my_API_token))
first_page_content <- rawToChar(first_page_get$content)
first_page_response <- fromJSON(first_page_content, flatten = TRUE)
first_next_page <- first_page_response$result$nextPage
first_page_elements <- first_page_response$result$elements

##### Get second page of 500 (1000)
second_page_get <- GET(first_next_page, add_headers("x-api-token" = my_API_token))
second_page_content <- rawToChar(second_page_get$content)
second_page_response <- fromJSON(second_page_content, flatten = TRUE)
second_next_page <- second_page_response$result$nextPage
second_page_elements <- second_page_response$result$elements

. . .

#### Get last page (k) of 500 contacts
k_page_get <- GET(`k-1`_next_page, add_headers("x-api-token" = my_API_token))
k_page_content <- rawToChar(k_page_get$content)
k_page_response <- fromJSON(k_page_content, flatten = TRUE)
k_next_page <- k_page_response$result$nextPage ## Should be NULL
k_page_elements <- k_page_response$result$elements

contacts_df <- bind_rows(first_page_elements, second_page_elements, third_page_elements, fourth_page_elements, fifth_page_elements, sixth_page_elements, seventh_page_elements, eighth_page_elements)
```

The code above only provides the basic information about a contact--ConctactID, first and last name, email address, phone, external data reference--the default variables Qualtrics has. There is no embedded data being pulled in from the code above. However, before I get to embedded data, I'd first like to do run the basic XM Directory API calls in a slightly more packaged way. 

Second, I'd also like to have the capability to extract embedded data of individual contacts within the directory for each contact. The current API call that includes embedded data will only get one contact's. I would like to see if extracting and formatting the embedded data call from a sample of the Directory or the entire Directory would be possible. These individual calls are done through calling a specific contact via their ContactID from Qualtrics one at a time. So, my current thoughts are that I would have to use the ContactID column from ```contacts_df``` to loop through a call to pull, then eventually extract and clean the embedded data from those in the Directory I wanted to explore in more sophisticated ways currently unavailable within Qualtrics.

I appreciate your willingness to consider additional functions.

**Comment by dsen6644** [comment]
I think it's possible to access mailing lists with a basic account as well.

The main hurdle is that the `append_root_url` function is specific to surveys and response exports. There is some opportunity to expand the package to include the ability to fetch information around mailing lists and distributions.

Here is a basic example of fetching the contacts for a standard contact list (without using `append_root_url`). The URL element would have to be expanded to include pools/directories associated with XM accounts.

```
fetch_mailing_list <- function(mailingListID){

  url <- paste0("https://",
                Sys.getenv("QUALTRICS_SUBDOMAIN"),
                ".qualtrics.com/API/v3/mailinglists/",
                mailingListID,
                "/contacts")

  ml <- list()

  while(!is.null(url)){

    x <- qualtrics_api_request("GET", url = url)
    elements <- x$result$elements
    url <- x$result$nextPage
    ml <- append(ml, elements)

  }

  ml <- ml %>%
    purrr::map(purrr::flatten) %>%
    purrr::modify_depth(1, ~ purrr::keep(., ~ !is.list(.x))) %>%
    dplyr::bind_rows()

  return(ml)

}
```
@juliasilge willing to discuss further around potential future functions like `fetch_mailing_list`, `fetch_distriubtion`, `all_distributions`, `fetch_distribution_history`. The final would mimic the output of "Download History". To my knowledge, no (current or deprecated) R package has had these features.

**Comment by juliasilge** [comment]
With the move to the updated API endpoint, we changed some of the internal functions for building URLs (`create_root_url()`, for example) and it should now be easier to create functions to access these other endpoints. Check out [`column_map()`](https://github.com/ropensci/qualtRics/blob/master/R/column_map.R) for an example of how it would look. A PR would be welcome to extend this package to include functionality for mailing lists, etc! ✨

**Comment by juliasilge** [comment]
Closed in #140
